optaplanner-examples/src/main/java/org/optaplanner/examples/pas/solver/score/PatientAdmissionMoveConstraintProvider.java

```diff
@@ -0,0 +1,180 @@
+                departmentMaximumAgeConstraint(constraintFactory),
+                requiredPatientEquipmentConstraint(constraintFactory),
+                assignEveryPatientToABedConstraint(constraintFactory),
+                preferredMaximumRoomCapacityConstraint(constraintFactory),
+                departmentSpecialismConstraint(constraintFactory),
+                roomSpecialismNotExistsConstraint(constraintFactory),
+                roomSpecialismNotFirstPriorityConstraint(constraintFactory),
+                preferredPatientEquipmentConstraint(constraintFactory)
+        };
+    }
+
+    public Constraint sameBedInSameNightConstraint(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUniquePair(BedDesignation.class)
+                .filter((left, right) -> left.getBed() != null
+                        && right.getBed() != null
+                        && left.getBed() == right.getBed()
+                        && left.getAdmissionPart().calculateSameNightCount(right.getAdmissionPart()) > 0)
+                .penalize("sameBedInSameNight", HardMediumSoftScore.ofHard(1000),
+                        (leftBd, rightBd) -> leftBd.getAdmissionPart().calculateSameNightCount(rightBd.getAdmissionPart()));
+    }
+
+    public Constraint femaleInMaleRoomConstraint(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BedDesignation.class)
+                .filter(bd -> bd.getPatientGender() == Gender.FEMALE
+                        && bd.getRoomGenderLimitation() == GenderLimitation.MALE_ONLY)
+                .penalize("femaleInMaleRoom", HardMediumSoftScore.ofHard(50), BedDesignation::getAdmissionPartNightCount);
+    }
+
+    public Constraint maleInFemaleRoomConstraint(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BedDesignation.class)
+                .filter(bd -> bd.getPatientGender() == Gender.MALE
+                        && bd.getRoomGenderLimitation() == GenderLimitation.FEMALE_ONLY)
+                .penalize("maleInFemaleRoom", HardMediumSoftScore.ofHard(50), BedDesignation::getAdmissionPartNightCount);
+    }
+
+    public Constraint differentGenderInSameGenderRoomInSameNightConstraint(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BedDesignation.class)
+                .filter(bd -> bd.getRoomGenderLimitation() == GenderLimitation.SAME_GENDER
+                        && bd.getBed() != null)
+                .join(BedDesignation.class,
+                        equal(BedDesignation::getRoom),
+                        lessThan(BedDesignation::getId),
+                        filtering((left, right) -> right.getRoomGenderLimitation() == GenderLimitation.SAME_GENDER
+                                && left.getAdmissionPart().calculateSameNightCount(right.getAdmissionPart()) > 0
+                                && left.getPatient().getGender() != right.getPatient().getGender()))
```

Comment:
Personally, I'd switch the order of operations to first do the `==`, then `!=` and only then do the `calculateSameNightCount(...)`. This way, you make sure that the (possibly expensive) calculation is only done when the quick checks are already true, never wasting time.

Reply:
https://github.com/kiegroup/optaplanner/pull/872/commits/7dd155181c6f1f7f867667d82eb53de3aa1d5a38